Implement applications to respond to ICMPv4 and ICMPv6 echo requests

aperezdc opened the pull request:
:warning: **WIP - I am working in test cases - do not merge just yet** :warning: 

This will allow pinging the AFTR.

kbara commented on src/apps/lwaftr/icmp.lua:
```diff
@@ -93,6 +94,22 @@ function new_icmpv4_packet(from_eth, to_eth, from_ip, to_ip, initial_pkt, l2_siz
 end
 
+function is_icmpv4(l2_size, pkt)
+   assert(l2_size > 2)
```
It is absolutely unacceptable to use assert here, other than as a debug aid. Assert in a situation like this cannot ever be merged. Bringing down the whole lwaftr rather than returning false is not ok. Asserts can be done in self tests and basically nowhere else, aside from very temporarily in branches that are not ready to merge into our main dev branch.

Also, we've switched back to using the the ethernet header size again rather than l2_size in this type of context, thanks to @wingo 's vlan app work.

aperezdc replied:
Fixed, I have removed all the `l2_size` thingies, and now the apps assume a normal Ethernet header.

kbara commented on src/apps/lwaftr/icmp.lua:
```diff
@@ -93,6 +94,22 @@ function new_icmpv4_packet(from_eth, to_eth, from_ip, to_ip, initial_pkt, l2_siz
 
+function is_icmpv4(l2_size, pkt)
+   assert(l2_size > 2)
+   return rd16(pkt.data + l2_size - 2) == constants.ethertype_ipv4
```
Constants.o_ethernet_ethertype
Easy to grep for, more readable, changeable if we ever back away from the current approach, less magic-number-y.

aperezdc replied:
The `assert()` is gone.

kbara commented on src/apps/lwaftr/icmp.lua:
```diff
@@ -93,6 +94,22 @@ function new_icmpv4_packet(from_eth, to_eth, from_ip, to_ip, initial_pkt, l2_siz
+   if is_icmpv4(l2_size, pkt) then
+      local ihl = get_ihl_from_offset(pkt, l2_size + constants.o_ipv4_ver_and_ihl)
+      return pkt.data[l2_size + ihl + constants.o_icmpv4_msg_type] == msg_type
+         and pkt.data[l2_size + ihl + constants.o_icmpv4_msg_code] == (msg_code or 0)
```
Why the magic with nil msg_codes?

kbara commented:
Looks like a good start. Please add an end-to-end test for each app (and a corresponding vlan one, even though that's a little tedious; Diego's .sh script makes it not too bad).

aperezdc commented:
I have fixed the nits in the updated PR, and added the missing `apps.lwaftr.icmp.is_icmv6_message()` function. I will update the PR in a while with a commit to wire the `ICMPEcho` applications in `setup.lua`, and another with the tests.

aperezdc commented:
While cooking the commit to wire things into `setup.lua`, I noticed that it is more convenient to have the apps to behave in the same way as `nd_light`, for example, with two “north” and “south” duplex ports. Right now only one of the ports is duplex, which is not so convenient to connect it to other apps. I will change that as well.

aperezdc commented:
Current version is working and has a simple end-to-end test. I have tried locally to generate a VLAN version of the test (using `add-vlan.sh` with the new pcaps), but the test case fails: packets are arriving to the `ICMPEcho` applications with the VLAN tags. I'll look more into this tomorrow.

kbara commented:
Thanks. :)

aperezdc commented:
Updated the last commit with the VLAN tests. The simple “just one ping” testcase is working for `end-to-end-vlan.sh` now as well. Next I'm adding test cases which have both packets to encapsulate/decapsulate _and_ a ping to the AFTR.

dpino commented on src/apps/lwaftr/icmp.lua:
```diff
@@ -93,6 +94,22 @@ function new_icmpv4_packet(from_eth, to_eth, from_ip, to_ip, initial_pkt, l2_siz
 end
 
+function is_icmpv4(pkt)
+   return rd16(pkt.data + constants.o_ethernet_ethertype) == constants.n_ethertype_ipv4
```
There's `is_ipv4` and `get_ipv4_proto`  defined at lwaftr.lua. At this moment these function are local. I prefer to reuse code as much as we can, so bugs are easier to fix (single point of failure). Ideally functions like this could be at `lwutil.lua` or other helper file. If it's easier just delocalize them and reuse them.

kbara replied:
I prefer leaving these local, and then moving them as soon as we actually concretely have a second caller in a different file actually re-using them. WDYT?

kbara replied:
Ah, misread. Yes, anything we're using in two different places should be re-used/shared.

dpino commented on src/apps/lwaftr/ipv6_apps.lua:
```diff
@@ -147,3 +152,61 @@ function Fragmenter:push ()
+         addresses[ffi.string(v, 16)] = true
+      end
+   end
+   return setmetatable({addresses = addresses}, {__index = ICMPEcho})
```
`address` is a 32-bit cdata object, correct?  If that's the case it's possible to convert the 32-bit cdata to a Lua number and use it as an index, probably is cheaper to run `tonumber()` than `ffi.tostring()`.

OTOH, there's a `set` function in `dump.lua`, probably not the best place. Since you need to create a set, maybe you can move this function to `lwutils.lua` and reuse it. I think using the `set` function makes the code easier to read.

aperezdc replied:
Here `address` is a `char[4]` cdata object, which won't convert directly to a Lua numeric value, hence the explicit use of `rd32()`.

dpino commented on src/program/snabb_lwaftr/setup.lua:
```diff
@@ -47,10 +49,16 @@ function lwaftr_app(c, conf)
+   config.link(c, "lwaftr.v6 -> icmpechov6.north")
+
+   set_postprocessors(c, "icmpechov6.south", postprocessing_apps_v6)
+   set_postprocessors(c, "icmpechov4.south", postprocessing_apps_v4)
```
I would prefer if you use the preprocessors and postprocessors logic, otherwise is going to be confuse. Setting the preprocessors and postprocessors to the lwaftr should be the last step. I think you want something like this:

```
append(preprocessing_apps_v4, { name = "icmpechov4", input = "south", output = "north" })
prepend(postprocessing_apps_v4, { name = "icmpechov4", input = "south", output = "north" })
append(preprocessing_apps_v6, { name = "icmpechov6", input = "south", output = "north" })
prepend(postprocessing_apps_v6, { name = "icmpechov6", input = "south", output = "north" })

set_preprocessors(c, preprocessing_apps_v4, "lwaftr.v4")
set_preprocessors(c, preprocessing_apps_v6, "lwaftr.v6")
set_postprocessors(c, "lwaftr.v6", postprocessing_apps_v6)
set_postprocessors(c, "lwaftr.v4", postprocessing_apps_v4)
```

And probably there's a bug in the `set_preprocessors` and `set_postprocessors` methods as they assume the name of the `input` and `output` links is always _input_ and _output_. This change should fix it:

```
function set_preprocessors(c, apps, dst)
   assert(type(apps) == "table")
   link_apps(c, apps)
   local last_app, output = apps[#apps], "output"
   if type(last_app) == "table" then
      last_app, output = last_app["name"], last_app["output"]
   end
   config.link(c, ("%s.%s -> %s"):format(last_app, output, dst))
end
```

Similarly for `set_postprocessors`.

aperezdc replied:
Thanks for the suggestion. Just one note: the strings for `input` and `output` have to be swapped in the list of postprocessing applications (traffic is flowing in the opposite direction there).

dpino commented on src/program/snabb_lwaftr/tests/data/add-vlan.sh:
```diff
@@ -77,6 +79,8 @@ V6=(
     udp-afteraftr-reassembled-ipv6-2frags.pcap
     udp-fromb4-2frags-bound.pcap
+	ping-v6.pcap
+	ping-v6-reply.pcap
```
Nit: I think there's tabs. Tip: turn on git's precommit hook.

wingo commented:
You will need to rebase this onto lwaftr_starfruit.  Be sure to not pull in extraneous commits.  Sorry for the noise and thanks!

aperezdc commented:
I have filed PR #220 which is the same as this, rebased on top of `lwaftr_starfruit`. That will be the one eventually getting merged, so I am closing this PR.
